adds a new OSD scenario to create OSDs with ceph-volume

andrewschoen opened the pull request:
``ceph-volume`` has not been released yet, so the testing scenario I've added here can not be enabled in the CI yet.  However, you can test this using the ``wip-volume`` branch of ceph from shaman.ceph.com. I've test that here and it's passing: https://jenkins.ceph.com/view/ceph-ansible/job/ceph-ansible-scenario/137/

leseb commented on roles/ceph-osd/tasks/main.yml:
```diff
@@ -27,6 +27,21 @@
   # Hard code this so we will skip the entire file instead of individual tasks (Default isn't Consistent)
   static: False
 
+- name: create lvm OSDs with ceph-volume
```
To stay consistent with the playbook, please do not use any capital letters.
Also, the other includes don't have names, I'm not saying it's a good/bad idea I'm thinking about the consistency here :)

andrewschoen replied:
I'll fix the capital letters.

Actually, when reading logs from QE I've wished that includes had a name because it's sometimes hard to tell what file was being 'skipped' when the include isn't needed. I'd be +1 to requiring names for any includes going forward.

leseb replied:
+1 on my side too.

leseb commented on roles/ceph-osd/tasks/main.yml:
```diff
@@ -27,6 +27,21 @@
+  # Hard code this so we will skip the entire file instead of individual tasks (Default isn't Consistent)
+  static: False
+
+- name: ensure OSD daemons are started
```
ditto

leseb commented on roles/ceph-osd/tasks/scenarios/lvm.yml:
```diff
@@ -0,0 +1,5 @@
+---
+
```
please remove the line here

leseb commented on roles/ceph-osd/tasks/scenarios/lvm.yml:
```diff
@@ -0,0 +1,5 @@
+---
+
+- name: use ceph-volume to create filestore OSDs with dedicated journals
```
ditto

leseb commented on roles/ceph-osd/tasks/start_osds.yml:
```diff
@@ -0,0 +1,19 @@
+---
+
```
please remove the line here

leseb commented on roles/ceph-osd/tasks/scenarios/lvm.yml:
```diff
@@ -0,0 +1,5 @@
+---
+
+- name: use ceph-volume to create filestore OSDs with dedicated journals
+  command: "ceph-volume lvm create --filestore --data {{ item.key }} --journal {{ item.value }}"
```
is the first implementation only working on filestore?

andrewschoen replied:
yes, the first implementation only works on filestore

leseb commented on roles/ceph-osd/tasks/check_mandatory_vars.yml:
```diff
@@ -55,6 +56,26 @@
     - not osd_auto_discovery
     - devices|length == 0
 
+- name: verify lvm_volumes have been provided
```
add another one to make sure `osd_objectstore == 'filestore'` only? If not, we fail.

andrewschoen replied:
yeah, I can do that

andrewschoen commented:
jenkins test this please

andrewschoen commented:
jenkins test this please

andrewschoen commented:
jenkins test jewel-ansible2.2-docker_cluster

andrewschoen commented:
jenkins test jewel-ansible2.2-purge_dmcrypt

andrewschoen commented:
jenkins test luminous-ansible2.2-docker_dedicated_journal

andrewschoen commented:
jenkins test luminous-ansible2.2-update_cluster

andrewschoen commented:
jenkins test luminous-ansible2.2-update_cluster

andrewschoen commented:
jenkins test luminous-ansible2.2-purge_cluster

andrewschoen commented:
jenkins test jewel-ansible2.2-centos7_cluster

andrewschoen commented:
jenkins test jewel-ansible2.2-purge_dmcrypt

leseb commented on roles/ceph-osd/tasks/check_mandatory_vars.yml:
```diff
@@ -30,6 +30,7 @@
     - not dmcrypt_dedicated_journal
+    - not lvm_osds
 
 - name: verify only one osd scenario was chosen
```
I think this needs to be updated too to reflect `lvm_osds`

andrewschoen replied:
While I understand the intent behind this task I think in practice it's gonna be very hard to maintain. This check already does not include all of the existing osd scenario combinations. YAML/ansible is just not a great at providing this kind of input validation, is there a better way we can do this?

leseb replied:
Agreed, hence: https://github.com/ceph/ceph-ansible/pull/1725

leseb commented on roles/ceph-osd/tasks/check_mandatory_vars.yml:
```diff
@@ -55,6 +56,36 @@
+    - not osd_auto_discovery
+    - lvm_volumes|length == 0
+
+- name: verify osd_objectstore is 'filestore' when using lvm_osds
```
I'll just have this task before `verify lvm_volumes have been provided`so we stop right away. There is no need to evaluate `verify lvm_volumes have been provided` to fail after because of a wrong object store.

leseb commented on roles/ceph-osd/tasks/check_mandatory_vars.yml:
```diff
@@ -96,3 +127,13 @@
     - osd_objectstore == 'bluestore'
     - ceph_release_num.{{ ceph_release }} < ceph_release_num.luminous
+
+- name: check if lvm_osds is supported by the selected ceph version
```
ditto, we should move it up.

leseb commented on roles/ceph-osd/tasks/scenarios/lvm.yml:
```diff
@@ -0,0 +1,4 @@
+---
+- name: use ceph-volume to create filestore OSDs with dedicated journals
```
no capital :)

alfredodeza commented on docs/source/osds/scenarios.rst:
```diff
@@ -0,0 +1,26 @@
+is only available when the ceph release is Luminous or greater.
+
+.. note::
+   The creation of the logical volumes is not supported by ceph-ansible, ceph-volume
```
``` ``ceph-volume`` ``` and  ``` ``ceph-ansible`` ``` (with double ticks all the time) so that it is consistent everywhere it is used

alfredodeza commented on roles/ceph-osd/tasks/check_mandatory_vars.yml:
```diff
@@ -46,6 +46,46 @@
 
+- name: check if lvm_osds is supported by the selected ceph version
+  fail:
+    msg: "lvm_osds is not supported by the selected Ceph version, use Luminous or above."
```
"Luminous or newer" reads better I think

andrewschoen replied:
Ok, I'll update

alfredodeza commented on docs/source/osds/scenarios.rst:
```diff
@@ -0,0 +1,26 @@
+lvm
+---
+This OSD scenario uses ``ceph-volume`` to create OSDs from logical volumes and
+is only available when the ceph release is Luminous or greater.
```
yeah same here, "Luminous or newer" I think would be consistent and since you aren't using a number, "greater" doesn't correlate to the named-version

alfredodeza commented on roles/ceph-osd/tasks/scenarios/lvm.yml:
```diff
@@ -0,0 +1,4 @@
+---
+- name: use ceph-volume to create filestore osds with dedicated journals
```
are all tasks in ceph-ansible excuted by root? all ceph-volume calls will need to be

andrewschoen replied:
yes, they are

alfredodeza commented on roles/ceph-osd/tasks/start_osds.yml:
```diff
@@ -0,0 +1,18 @@
+  always_run: true
+  register: osd_id
+
+- name: start and add that the osd service(s) to the init sequence
```
I am not sure how this is going to work here. This would only work if ceph-volume has activated the OSD. The activation process is the one in charge of making sure that everything an OSD needs to start is there *including* starting the OSD process at the end.

In what situation one would want to start the OSD daemon like this?

andrewschoen replied:
This name is a bit misleading because it does not enable the systemd unit anymore, it's just ensuring that it is started. If we deploy and ceph-volume does not start the OSD this task will fail and notify the user.

alfredodeza commented on roles/ceph-osd/defaults/main.yml:
```diff
@@ -187,6 +188,18 @@ dedicated_devices: []
+# III. Use ceph-volume to create OSDs from logical volumes.
+# Use 'osd_scenario: lvm' to enable this scenario. Currently we only support dedicated journals
+# when using lvm, not collocated journals.
+# lvm_volumes is a dictionary whose key/value pair represent a data lv and a journal pair.
```
can we update this comment (and the docs) to mention that any LVM value (like an lv or vg) has to be a *name* and not a path?

andrewschoen replied:
Yeah, I can update that.

andrewschoen commented:
jenkins test luminous-ansible2.2-bluestore_cluster

andrewschoen commented:
jenkins test luminous-ansible2.2-centos7_cluster

andrewschoen commented:
jenkins test luminous-ansible2.2-docker_dmcrypt_journal_collocation

andrewschoen commented:
jenkins test luminous-ansible2.2-update_docker_cluster

andrewschoen commented:
jenkins test luminous-ansible2.2-bluestore_docker_dmcrypt_journal_collocation

andrewschoen commented:
I've got a successful test of the new ``lvm_osds`` scenario here, using the ``wip-volume`` branch of ceph. https://jenkins.ceph.com/view/ceph-ansible/job/ceph-ansible-scenario/150/console

Here's the part of the log where ``ceph-volume`` creates the osd:

```
TASK [ceph-osd : use ceph-volume to create filestore osds with dedicated journals] ***
task path: /home/jenkins-build/build/workspace/ceph-ansible-scenario/roles/ceph-osd/tasks/scenarios/lvm.yml:2
changed: [osd0] => (item={'key': u'test_volume', 'value': u'/dev/sdb'}) => {
    "changed": true, 
    "cmd": [
        "ceph-volume", 
        "lvm", 
        "create", 
        "--filestore", 
        "--data", 
        "test_volume", 
        "--journal", 
        "/dev/sdb"
    ], 
    "delta": "0:00:18.427975", 
    "end": "2017-08-04 11:26:49.833672", 
    "item": {
        "key": "test_volume", 
        "value": "/dev/sdb"
    }, 
    "rc": 0, 
    "start": "2017-08-04 11:26:31.405697", 
    "warnings": []
}

STDOUT:

Running command: ceph-authtool --gen-print-key
Running command: ceph --cluster ceph --name client.bootstrap-osd --keyring /var/lib/ceph/bootstrap-osd/ceph.keyring -i - osd new 95fec51d-719a-4d08-bf55-c71ac94c5043
Running command: sudo vgs --reportformat=json
Running command: sudo lvs -o lv_tags,lv_path,lv_name,vg_name --reportformat=json
Running command: sudo vgs --reportformat=json
Running command: sudo lvs -o lv_tags,lv_path,lv_name,vg_name --reportformat=json
Running command: sudo lvchange --addtag ceph.osd_id=0 /dev/test_group/test_volume
Running command: sudo lvchange --addtag ceph.osd_fsid=95fec51d-719a-4d08-bf55-c71ac94c5043 /dev/test_group/test_volume
Running command: sudo lvchange --addtag ceph.type=data /dev/test_group/test_volume
Running command: sudo lvchange --addtag ceph.cluster_fsid=2f4ad268-1440-4b49-9b84-d6f97655e738 /dev/test_group/test_volume
Running command: sudo lvchange --addtag ceph.journal_device=/dev/sdb /dev/test_group/test_volume
Running command: sudo lvchange --addtag ceph.data_device=/dev/test_group/test_volume /dev/test_group/test_volume
Running command: sudo lvs -o lv_tags,lv_path,lv_name,vg_name --reportformat=json
Running command: ceph-authtool --gen-print-key
Running command: sudo mkfs -t xfs -f -i size=2048 /dev/test_group/test_volume
 stdout: meta-data=/dev/test_group/test_volume isize=2048   agcount=4, agsize=3276544 blks
         =                       sectsz=512   attr=2, projid32bit=1
         =                       crc=1        finobt=0, sparse=0
data     =                       bsize=4096   blocks=13106176, imaxpct=25
         =                       sunit=0      swidth=0 blks
naming   =version 2              bsize=4096   ascii-ci=0 ftype=1
log      =internal log           bsize=4096   blocks=6399, version=2
         =                       sectsz=512   sunit=0 blks, lazy-count=1
realtime =none                   extsz=4096   blocks=0, rtextents=0
Running command: sudo mount -t xfs -o noatime,inode64 /dev/test_group/test_volume /var/lib/ceph/osd/ceph-0
Running command: sudo ln -s /dev/sdb /var/lib/ceph/osd/ceph-0/journal
Running command: sudo ceph --cluster ceph --name client.bootstrap-osd --keyring /var/lib/ceph/bootstrap-osd/ceph.keyring mon getmap -o /var/lib/ceph/osd/ceph-0/activate.monmap
 stderr: got monmap epoch 1
Running command: chown -R ceph:ceph /dev/sdb
Running command: chown -R ceph:ceph /var/lib/ceph/osd/ceph-0/
Running command: sudo ceph-osd --cluster ceph --mkfs -i 0 --monmap /var/lib/ceph/osd/ceph-0/activate.monmap --osd-data /var/lib/ceph/osd/ceph-0/ --osd-journal /var/lib/ceph/osd/ceph-0/journal --osd-uuid 95fec51d-719a-4d08-bf55-c71ac94c5043 --setuser ceph --setgroup ceph
 stderr: 2017-08-04 11:26:43.953089 7fbb30058d00 -1 journal check: ondisk fsid 00000000-0000-0000-0000-000000000000 doesn't match expected 95fec51d-719a-4d08-bf55-c71ac94c5043, invalid (someone else's?) journal
 stderr: 2017-08-04 11:26:44.313094 7fbb30058d00 -1 read_settings error reading settings: (2) No such file or directory
Running command: ceph-authtool /var/lib/ceph/osd/ceph-0/keyring --create-keyring --name osd.0 --add-key AQDnWYRZaA+SHBAALkbgD0T9nYPpXj8ovWtRcw==
 stdout: creating /var/lib/ceph/osd/ceph-0/keyring
added entity osd.0 auth auth(auid = 18446744073709551615 key=AQDnWYRZaA+SHBAALkbgD0T9nYPpXj8ovWtRcw== with 0 caps)
Running command: chown -R ceph:ceph /var/lib/ceph/osd/ceph-0/keyring
Running command: sudo lvs -o lv_tags,lv_path,lv_name,vg_name --reportformat=json
Running command: sudo mount -v /dev/test_group/test_volume /var/lib/ceph/osd/ceph-0
 stderr: mount: /dev/mapper/test_group-test_volume is already mounted or /var/lib/ceph/osd/ceph-0 busy
 stderr: /dev/mapper/test_group-test_volume is already mounted on /var/lib/ceph/osd/ceph-0
Running command: chown -R ceph:ceph /dev/sdb
Running command: sudo systemctl enable ceph-volume@lvm-0-95fec51d-719a-4d08-bf55-c71ac94c5043
 stderr: Created symlink from /etc/systemd/system/multi-user.target.wants/ceph-volume@lvm-0-95fec51d-719a-4d08-bf55-c71ac94c5043.service to /usr/lib/systemd/system/ceph-volume@.service.
Running command: sudo systemctl start ceph-osd@0
```
